src/gs-update-monitor.c

```diff
@@ -48,10 +54,46 @@ struct _GsUpdateMonitor {
 	guint		 check_hourly_id;		/* and then every hour */
 	guint		 check_daily_id;		/* every 3rd day */
 	guint		 notification_blocked_id;	/* rate limit notifications */
+
+	MwscScheduler	*scheduler;
+	GHashTable	*scheduled_updates; /* (element-type utf8 UpdateScheduleHelper) */
+	GCancellable	*scheduled_updates_cancellable;
 };
 
 G_DEFINE_TYPE (GsUpdateMonitor, gs_update_monitor, G_TYPE_OBJECT)
 
+typedef struct {
+	GsUpdateMonitor *monitor;
+	MwscScheduleEntry *entry;
+	GsApp *app;
+	gulong download_now_handler_id;
+	gulong invalidate_handler_id;
+} UpdateScheduleHelper;
+
+static void
+download_schedule_helper_free (UpdateScheduleHelper *helper)
+{
+	if (helper->entry != NULL) {
+		g_debug ("Unscheduling update for app %s, with entry id %s",
+			 gs_app_get_unique_id (helper->app),
+			 mwsc_schedule_entry_get_id (helper->entry));
+
+		if (helper->download_now_handler_id > 0)
+			g_signal_handler_disconnect (helper->entry,
+						     helper->download_now_handler_id);
+		if (helper->invalidate_handler_id > 0)
+			g_signal_handler_disconnect (helper->entry,
+						     helper->invalidate_handler_id);
+		mwsc_schedule_entry_remove_async (helper->entry, NULL, NULL, NULL);
```

Comment:
I’m a little uneasy about firing off an async function like this without tracking whether it completes. Does the relevant `GMainContext` ever get iterated? Mogwai will automatically remove its schedule entries if gnome-software drops off D-Bus (i.e. when it exits), but not before then.

Reply:
I don't understand. I am not interested in the result of this function as, if it fails, there's little value in printing a warning. But if there's something I am missing, I can add the callback...

Reply:
You might not be interested in the result, but you should be interested in whether the call has completed.